### PR TITLE
First version.

### PR DESCRIPTION
**Pull request by azubieta**

### CODE REVIEW

File: include/elfutil/ElfFile.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// system includes
```
Comment (TheAssassin) [inline include/elfutil/ElfFile.h]: I'd prefer lower case public header filenames, like the rest of the world uses them (except Qt, maybe).

File: include/elfutil/errors.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        explicit ElfFileParseError(const std::string& msg);
+    };
+
+    // thrown by traceDynamicDependencies() if a dependency is missing
+    class DependencyNotFoundError : public std::runtime_error {
+    public:
+        explicit DependencyNotFoundError(const std::string& msg);
+    };
+
+}
```
Comment (TheAssassin) [inline include/elfutil/errors.h]: Missing trailing newline.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+add_library(elfutil SHARED
+    elfutil/ElfFile.cpp
+    elfutil/log.cpp
+    elfutil/errors.cpp)
```
Comment (TheAssassin) [inline src/CMakeLists.txt]:
```
add_library(elfutil SHARED
￼     elfutil/ElfFile.cpp
￼     elfutil/log.cpp
￼     elfutil/errors.cpp
)
```

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+target_include_directories(elfutil
+    PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}
+    )
```
Comment (TheAssassin) [inline src/CMakeLists.txt]:
```suggestion
)
```

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+set_target_properties(elfutil PROPERTIES
+    VERSION ${libelfutil_VERSION} SOVERSION ${libelfutil_VERSION_MAJOR}
+    PUBLIC_HEADER
```
Comment (TheAssassin) [inline src/CMakeLists.txt]: Please use a single line.

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    elfutil/log.cpp
+    elfutil/errors.cpp)
+
+target_include_directories(elfutil
+    PUBLIC $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}
+    )
+
+set_target_properties(elfutil PROPERTIES
+    VERSION ${libelfutil_VERSION} SOVERSION ${libelfutil_VERSION_MAJOR}
```
Comment (TheAssassin) [inline src/CMakeLists.txt:18]: Is `SOVERSION` standard? Do we have to introduce that elsewhere? From which version of CMake is it available?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    PUBLIC_HEADER
+        "${PROJECT_SOURCE_DIR}/include/elfutil/ElfFile.h;${PROJECT_SOURCE_DIR}/include/elfutil/errors.h;")
+
+target_link_libraries(elfutil PUBLIC ${Boost_LIBRARIES})
+
+install(TARGETS elfutil
+    EXPORT ElfUtilTargets
+    LIBRARY DESTINATION "${INSTALL_LIB_DIR}" COMPONENT lib
+    PUBLIC_HEADER DESTINATION "${INSTALL_INCLUDE_DIR}/elfutil" COMPONENT dev)
```
Comment (TheAssassin) [inline src/CMakeLists.txt]:
```
install(
    TARGETS elfutil
￼    EXPORT ElfUtilTargets
￼    LIBRARY DESTINATION "${INSTALL_LIB_DIR}" COMPONENT lib
￼    PUBLIC_HEADER DESTINATION "${INSTALL_INCLUDE_DIR}/elfutil" COMPONENT dev
)

File: src/elfutil/errors.cpp [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#include "elfutil/errors.h"
+
+namespace elfutil {
+
+    ElfFileParseError::ElfFileParseError(const std::string& msg) : std::runtime_error(msg) {}
+
+    DependencyNotFoundError::DependencyNotFoundError(const std::string& msg) : std::runtime_error(msg) {}
+}
```
Comment (TheAssassin) [inline src/elfutil/errors.cpp]: Missing trailing newline.

File: src/elfutil/log.cpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    prependSpace = false;
+                    break;
+                default:
+                    break;
+            }
+
+            return ldLog(prependSpace, logLevelSet, currentLogLevel);
+        }
+    }
+}
```
Comment (TheAssassin) [inline src/elfutil/log.cpp:126]: I guess we don't need a copy of linuxdeploy's entire logging module in here. In fact, I'd remove any sort of logging from the library (except for debugging, maybe, but something simpler should do for that purpose).

File: src/elfutil/misc.hpp [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            std::vector<char> buf(PATH_MAX, '\0');
+
+            if (readlink("/proc/self/exe", buf.data(), buf.size()) < 0) {
+                return "";
+            }
+
+            return buf.data();
+        }
+    }
+}
```
Comment (TheAssassin) [inline src/elfutil/misc.hpp:68]: We don't need half of these methods. Please remove anything we don't need.

File: tests/simple_library/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,2 @@
+add_library(simple_library SHARED simple_library.cpp)
+target_link_libraries(simple_library)
```
Comment (TheAssassin) [inline tests/simple_library/CMakeLists.txt]: Missing trailing newline.

**Review by TheAssassin — CHANGES_REQUESTED** [review]
It doesn't make any sense to manage one half of the dependencies using submodules, the other one using this DownloadProject script. Please switch everything to submodules.

As in many of your PRs, there's lots of files lacking a trailing newline character. The GitHub browser reveals that. Please either take care of it manually, or find a technical way to automate this (IDEs usually have an option for that).

File: src/elfutil/log.cpp (hunk 9 above)
Comment (azubieta) [reply src/elfutil/log.cpp:126]: I'll create an issue for this.

File: src/CMakeLists.txt (hunk 6 above)
Comment (azubieta) [reply src/CMakeLists.txt:18]: https://cmake.org/cmake/help/v3.0/prop_tgt/SOVERSION.html

**Comment by azubieta** [comment]
All issues fixed, proceeding to merge.

**Pull request merged**